jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java

```diff
@@ -382,6 +382,10 @@ public Runnable onIdleTimeout(TimeoutException t)
 
                 // Otherwise We ask any idle timeout listeners if we should call onFailure or not
                 onIdleTimeout = _onIdleTimeout;
+
+                // If there is no idle nor failure listener, then we can fail the callback directly without double lock
+                if (onIdleTimeout == null && _onFailure == null && _request != null)
+                    return () -> _request._callback.failed(t);
```

Comment:
> Just noting that there is some things `onFailure` would do that aren't done in this optimisation case.
> 
> * making a temp request for logging, and producing an error response if one hasn't been created.
That is only done if the _request is null, hence the not null check at 387

> * setting _failure content chunk.

That is only needed if something is to read content, but there has been no notification to the app that there is something to read and the callback is failed anyway.

So I think this is fine.... but I might set the _failure chunk as well, just in case.

Reply:
@lachlan-roberts I refined it a bit so that we always set `_failure` if we are going to fail the callback.